File: app/src/main/java/aeropresscipe/divinelink/aeropress/generaterecipe/GenerateRecipeFragment.java
package aeropresscipe.divinelink.aeropress.generaterecipe;

import android.os.Bundle;

import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Handler;
import android.os.Looper;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.Button;
import android.widget.LinearLayout;


import aeropresscipe.divinelink.aeropress.R;
import aeropresscipe.divinelink.aeropress.customviews.Notification;
import aeropresscipe.divinelink.aeropress.timer.TimerActivity;

public class GenerateRecipeFragment extends Fragment implements GenerateRecipeView {

    private RecyclerView recipeRv;
    private LinearLayout generateRecipeButton;
    private LinearLayout timerButton;
    private Button resumeBrewBtn;

    private Animation mFadeInAnimation;
    private Animation mAdapterAnimation;
    private GenerateRecipePresenter presenter;

    private Recipe recipe;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.fragment_generate_recipe, container, false);

        recipeRv = v.findViewById(R.id.recipe_rv);
        generateRecipeButton = v.findViewById(R.id.generateRecipeButton);
        timerButton = v.findViewById(R.id.startTimerButton);
        resumeBrewBtn = v.findViewById(R.id.resumeBrewButton);


        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        recipeRv.setLayoutManager(layoutManager);

        presenter = new GenerateRecipePresenterImpl(this);
        presenter.getRecipe(getContext());
        mFadeInAnimation = AnimationUtils.loadAnimation(getActivity(), R.anim.initiliaze_animation);

        initListeners();

        return v;
    }

    private void initListeners() {
        generateRecipeButton.setOnClickListener(view -> presenter.getNewRecipe(getContext(), false));

        generateRecipeButton.setOnLongClickListener(view -> {
            presenter.getNewRecipe(getContext(), true);
            return true;
        });

        timerButton.setOnClickListener(view -> {
            recipe.setNewRecipe(true);
            startActivity(TimerActivity.newIntent(requireContext(), recipe));
        });

        resumeBrewBtn.setOnClickListener(view -> {
            recipe.setNewRecipe(false);
            startActivity(TimerActivity.newIntent(requireContext(), recipe));
        });
    }

    @Override
    public void showRecipe(final Recipe randomRecipe) {
        if (getActivity() != null) {
            getActivity().runOnUiThread(new Runnable() {
                final GenerateRecipeRvAdapter recipeRvAdapter = new GenerateRecipeRvAdapter(randomRecipe, getActivity());
                @Override
                public void run() {
                    recipeRv.setAdapter(recipeRvAdapter);
                    if (!mFadeInAnimation.hasEnded()) {
                        try {
                            //FIXME Maybe this isn't the best solution. Basically, there's a problem where if you're using Fragment Transition animation, and you also want to load the following animation on the button, app crashes.
                            //So I used this try catch method and app waits until UI loads.
                            //This happens because, getContext() is null since there are parts of the lifecycle where this will return null.
                            mFadeInAnimation = AnimationUtils.loadAnimation(getContext(), R.anim.fade_in_out);
                        } catch (Exception ignore) {
                            // do nothing
                        }
                        resumeBrewBtn.startAnimation(mFadeInAnimation);
                    }
                }
            });
        }
    }

    @Override
    public void passData(Recipe recipe) {
        this.recipe = recipe;
    }

    @Override
    public void showIsAlreadyBrewingDialog() {
        Notification.Companion.make(generateRecipeButton, R.string.alreadyBrewingDialog).setAnchorView(R.id.bottom_navigation).show();
    }

    @Override
    public void showRecipeRemoveResume(final Recipe randomRecipe) {
        if (getActivity() != null) {
            getActivity().runOnUiThread(new Runnable() {
                final GenerateRecipeRvAdapter recipeRvAdapter = new GenerateRecipeRvAdapter(randomRecipe, getActivity());

                @Override
                public void run() {
                    mAdapterAnimation = AnimationUtils.loadAnimation(getActivity(), R.anim.adapter_anim);
                    recipeRv.startAnimation(mAdapterAnimation);
                    // We need this so the adapter changes during the animation phase, and not before it.
                    Handler adapterHandler = new Handler(Looper.getMainLooper());
                    Runnable adapterRunnable = () -> recipeRv.setAdapter(recipeRvAdapter);

                    adapterHandler.postDelayed(adapterRunnable, mAdapterAnimation.getDuration());

                    if (!mFadeInAnimation.hasEnded()) {
                        mFadeInAnimation = AnimationUtils.loadAnimation(getActivity(), R.anim.fade_out);
                        resumeBrewBtn.startAnimation(mFadeInAnimation);
                    }

                }
            });
        }
    }

    @Override
    public void showRecipeAppStarts(final Recipe randomRecipe) {
        if (getActivity() != null) {
            getActivity().runOnUiThread(new Runnable() {
                final GenerateRecipeRvAdapter recipeRvAdapter = new GenerateRecipeRvAdapter(randomRecipe, getActivity());
                @Override
                public void run() {
                    recipeRv.setAdapter(recipeRvAdapter);
                    if (mFadeInAnimation == null)
                        mFadeInAnimation = AnimationUtils.loadAnimation(getActivity(), R.anim.initiliaze_animation);
                    resumeBrewBtn.startAnimation(mFadeInAnimation);
                }
            });
        }
    }

    public static GenerateRecipeFragment newInstance() {
        Bundle args = new Bundle();
        GenerateRecipeFragment fragment = new GenerateRecipeFragment();
        fragment.setArguments(args);
        return fragment;
    }
}

File: app/src/main/java/aeropresscipe/divinelink/aeropress/generaterecipe/GenerateRecipeFragment.kt
package aeropresscipe.divinelink.aeropress.generaterecipe

import  aeropresscipe.divinelink.aeropress.R
import aeropresscipe.divinelink.aeropress.base.TimerViewCallback
import aeropresscipe.divinelink.aeropress.components.saverecipecard.SaveRecipeCardView.Companion.DISLIKE_MAX_FRAME
import aeropresscipe.divinelink.aeropress.components.saverecipecard.SaveRecipeCardView.Companion.DISLIKE_MIN_FRAME
import aeropresscipe.divinelink.aeropress.components.saverecipecard.SaveRecipeCardView.Companion.LIKE_MAX_FRAME
import aeropresscipe.divinelink.aeropress.components.saverecipecard.SaveRecipeCardView.Companion.LIKE_MIN_FRAME
import aeropresscipe.divinelink.aeropress.components.snackbar.Notification
import aeropresscipe.divinelink.aeropress.databinding.FragmentGenerateRecipeBinding
import aeropresscipe.divinelink.aeropress.helpers.LottieHelper
import aeropresscipe.divinelink.aeropress.home.HomeViewModel
import aeropresscipe.divinelink.aeropress.settings.SettingsActivity
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import androidx.annotation.Px
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.ViewModelProvider
import com.airbnb.lottie.LottieAnimationView
import dagger.hilt.android.AndroidEntryPoint
import gr.divinelink.core.util.extensions.padding
import gr.divinelink.core.util.extensions.updatePaddingAnimator
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import kotlinx.datetime.LocalDateTime
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toLocalDateTime
import timber.log.Timber
import java.lang.ref.WeakReference
import javax.inject.Inject

@AndroidEntryPoint
class GenerateRecipeFragment :
    Fragment(),
    IGenerateRecipeViewModel,
    GenerateRecipeStateHandler,
    TimerViewCallback {
    private var binding: FragmentGenerateRecipeBinding? = null

    @Inject
    lateinit var assistedFactory: GenerateRecipeViewModelAssistedFactory
    private lateinit var viewModel: GenerateRecipeViewModel
    private val parentViewModel: HomeViewModel by activityViewModels()

    private lateinit var fadeInAnimation: Animation
    private lateinit var adapterAnimation: Animation

    private val currentMoment: Instant by lazy { Clock.System.now() }
    private val datetime: LocalDateTime by lazy { currentMoment.toLocalDateTime(TimeZone.currentSystemDefault()) }

    private var lottieFavorite: LottieAnimationView? = null

    @Px
    private var coordinatorPadding: Int = 0

    var refresh: Boolean = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentGenerateRecipeBinding.inflate(inflater, container, false)
        val view = binding?.root

        val viewModelFactory = GenerateRecipeViewModelFactory(assistedFactory, WeakReference<IGenerateRecipeViewModel>(this))
        viewModel = ViewModelProvider(this, viewModelFactory)[GenerateRecipeViewModel::class.java]
        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Timber.d("onViewCreated")
        viewModel.init(datetime.hour)
        val favoriteMenuItem = binding?.toolbar?.menu?.findItem(R.id.favorite)
        favoriteMenuItem?.setActionView(R.layout.layout_favorite_item)
        lottieFavorite = favoriteMenuItem?.actionView as LottieAnimationView?

        LottieHelper.updateLikeButton(lottieFavorite)

        fadeInAnimation = AnimationUtils.loadAnimation(requireContext(), R.anim.initiliaze_animation)
        adapterAnimation = AnimationUtils.loadAnimation(requireContext(), R.anim.adapter_anim)
        binding?.coordinator?.padding(bottom = coordinatorPadding)
        initListeners()
    }

    override fun onResume() {
        super.onResume()
        Timber.d("Resume")
        viewModel.getRecipe(refresh)
    }

    override fun updateBottomPadding(bottomPadding: Int) {
        coordinatorPadding = bottomPadding
        binding?.coordinator?.updatePaddingAnimator(bottom = bottomPadding)
    }

    override fun updateState(state: GenerateRecipeState) {
        when (state) {
            is GenerateRecipeState.InitialState -> handleInitialState()
            is GenerateRecipeState.ErrorState -> handleErrorState(state)
            is GenerateRecipeState.LoadingState -> handleLoadingState()
            is GenerateRecipeState.ShowAlreadyBrewingState -> handleShowAlreadyBrewingState()
            is GenerateRecipeState.ShowRecipeState -> handleShowRecipeState(state)
            is GenerateRecipeState.RefreshRecipeState -> handleRefreshRecipeState(state)
            is GenerateRecipeState.StartTimerState -> handleStartTimerState(state)
            is GenerateRecipeState.HideResumeButtonState -> handleHideResumeButtonState()
            is GenerateRecipeState.UpdateToolbarState -> handleUpdateToolbarState(state)
            is GenerateRecipeState.ShowSnackBar -> handleShowSnackBar(state)
            is GenerateRecipeState.UpdateSavedIndicator -> handleUpdateSavedIndicator(state)
            is GenerateRecipeState.RecipeRemovedState -> handleRecipeRemovedState()
            is GenerateRecipeState.RecipeSavedState -> handleRecipeSavedState()
        }
    }

    override fun handleHideResumeButtonState() {
        parentViewModel.generateRecipe()
    }

    override fun handleInitialState() {
        // Intentionally Blank.
    }

    override fun handleLoadingState() {
        // Nothing yet
    }

    override fun handleErrorState(state: GenerateRecipeState.ErrorState) {
        // Nothing yet
    }

    override fun handleShowAlreadyBrewingState() {
        Notification.make(binding?.generateRecipeButton, R.string.alreadyBrewingDialog)
            .show()
    }

    override fun handleShowRecipeState(state: GenerateRecipeState.ShowRecipeState) {
        Timber.d("Recipe updated. Set refresh to false.")
        binding?.recipeList?.adapter = GenerateRecipeListView(
            steps = state.steps,
            context = requireContext()
        )
        refresh = false
    }

    override fun handleRefreshRecipeState(state: GenerateRecipeState.RefreshRecipeState) {
        binding?.recipeList?.startAnimation(adapterAnimation)
        MainScope().launch {
            delay(adapterAnimation.duration)
            binding?.recipeList?.adapter = GenerateRecipeListView(
                steps = state.steps,
                context = requireContext()
            )
        }
    }

    override fun handleStartTimerState(state: GenerateRecipeState.StartTimerState) {
        parentViewModel.startTimer(state.recipe, state.flow)
    }

    override fun handleUpdateToolbarState(state: GenerateRecipeState.UpdateToolbarState) {
        binding?.toolbar?.title = resources.getString(state.title)
    }

    override fun handleUpdateSavedIndicator(state: GenerateRecipeState.UpdateSavedIndicator) {
        lottieFavorite?.setMinAndMaxFrame(state.frame, state.frame)
        lottieFavorite?.playAnimation()
    }

    override fun handleShowSnackBar(state: GenerateRecipeState.ShowSnackBar) {
        Notification.make(binding?.generateRecipeButton, resources.getString(state.value.string, getString(state.value.favorites)))
            .show()
    }

    override fun handleRecipeSavedState() {
        lottieFavorite?.apply {
            setMinAndMaxFrame(LIKE_MIN_FRAME, LIKE_MAX_FRAME)
            playAnimation()
        }
    }

    override fun handleRecipeRemovedState() {
        lottieFavorite?.apply {
            setMinAndMaxFrame(DISLIKE_MIN_FRAME, DISLIKE_MAX_FRAME)
            playAnimation()
        }
    }

    private fun initListeners() {
        binding?.apply {
            generateRecipeButton.setOnClickListener {
                viewModel.generateRecipe()
            }
            generateRecipeButton.setOnLongClickListener {
                viewModel.forceGenerateRecipe()
                true
            }
            startTimerButton.setOnClickListener {
                viewModel.startTimer(resume = false)
            }
            resumeBrewButton.setOnClickListener {
                viewModel.startTimer(resume = true)
            }
        }
        lottieFavorite?.setOnClickListener {
            viewModel.likeRecipe()
        }

        binding?.toolbar?.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.menu_settings -> {
                    startActivity(Intent(context, SettingsActivity::class.java))
                    true
                }
                else -> false
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    companion object {
        @JvmStatic
        fun newInstance(): GenerateRecipeFragment {
            val args = Bundle()
            val fragment = GenerateRecipeFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
